[Fix] Throw BinderException for UNNEST expressions in WINDOW expressions

taniabogatsch opened the pull request:
Fixes https://github.com/duckdblabs/duckdb-internal/issues/1450.

We allow subqueries in `WINDOW` expressions, so the following query returns an expected result.
```sql
statement ok
CREATE TABLE tbl AS SELECT 42 AS i;

query I
SELECT SUM(i) OVER (ROWS BETWEEN (SELECT UNNEST([1])) PRECEDING AND 1 FOLLOWING) FROM tbl;
----
42
```

However, we do not allow `UNNEST` to be an expression outside a subquery in `WINDOW` expressions. Previously, queries of the following nature would throw an `InternalException`. This PR ensures that we now throw the expected binder exception.
```sql
statement ok
CREATE TABLE tbl AS SELECT 42 AS i;

statement error
SELECT SUM(i) OVER (ROWS BETWEEN UNNEST([1]) PRECEDING AND 1 FOLLOWING) FROM tbl;
----
UNNEST not supported here
```

### Relevant changes
- I moved `inside_window` from the `BaseSelectBinder` into the `ExpressionBinder`, to access it without type casting of binders in macro binding code-paths.
- I added two new functions to the `ExpressionBinder` to (1) detect `UNNEST` expressions (`WindowContainsUnnest`), and to (2) bind a `WINDOW` child expression with extra checks (`BindWindowChild`).
- I also did some code tidying, as my IDE complained about shadowed variables.

taniabogatsch committed:
unnest fixes in window expressions and shadowed variable renaming

taniabogatsch committed:
Merge branch 'main' into fix-fuzzer

taniabogatsch committed:
Merge branch 'main' into fix-fuzzer

Mytherin commented:
Thanks for the PR! 

I'm a bit worried about the added calls to `WindowContainsUnnest`. It seems to be adding a good amount of code complexity, and there are performance ramifications to (repeatedly) doing a full expression tree traversal during binding. Since we are already doing a full expression tree traversal during the actual bind this could lead to `O(n^2)` binding performance in the size of the expression tree. Of course often binding performance is not critical - but quadratic performance is always best avoided.

I've tried a simpler change that uses the existing `inside_window` property [here](https://github.com/Mytherin/duckdb/commit/62e3cfa629a88637856ccaa3ec2c8e52159e4820) and it seems to solve the issue as well (at least all of the tests seem to pass). Since you have looked at this more perhaps I'm missing something and this doesn't fully solve the problem?

taniabogatsch committed:
revert changes but keep tidy fixes

taniabogatsch commented:
@Mytherin, thanks for your review and the great suggestion! I've reverted the changes and added your change. I've only kept the other tidy changes. Let's see what CI has to say. :)

Mytherin committed:
Merge branch 'main' into fix-fuzzer

Mytherin commented:
Thanks!
